MapleServer2/PacketHandlers/Game/ShopHandler.cs

```diff
@@ -0,0 +1,137 @@
+                    HandleSell(session, packet);
+                    break;
+                default:
+                    IPacketHandler<GameSession>.LogUnknownMode(mode);
+                    break;
+            }
+        }
+
+        public static void HandleOpen(GameSession session, IFieldObject<Npc> npcFieldObject)
+        {
+            NpcMetadata metadata = NpcMetadataStorage.GetNpc(npcFieldObject.Value.Id);
+
+            ShopMetadata shop = ShopMetadataStorage.GetShop(metadata.ShopId);
+            if (shop == null)
+            {
+                return;
+            }
+
+            session.Send(ShopPacket.Open(shop));
+            session.Send(ShopPacket.LoadProducts(shop.Items));
+            session.Send(ShopPacket.Reload());
+            session.Send(NpcTalkPacket.Respond(npcFieldObject, NpcType.Default, DialogType.None, 0));
+        }
+
+        private static void HandleClose(GameSession session)
+        {
+            session.Send(ShopPacket.Close());
+        }
+
+        private static void HandleSell(GameSession session, PacketReader packet)
+        {
+            // sell to shop
+            long itemUid = packet.ReadLong();
+            int quantity = packet.ReadInt();
+
+            if (!session.Player.Inventory.Items.TryGetValue(itemUid, out Item item))
+            {
+                return;
+            }
+
+            int price = ItemMetadataStorage.GetCustomSellPrice(item.Id);
+            session.Player.Wallet.Meso.Modify(price * quantity);
+
+            InventoryController.Consume(session, item.Uid, quantity);
+
+            session.Send(ShopPacket.Sell(item.Id, quantity));
+        }
+
+        private static void HandleBuy(GameSession session, PacketReader packet)
+        {
+            int itemUid = packet.ReadInt();
+            int quantity = packet.ReadInt();
+
+            ShopItem shopItem = ShopMetadataStorage.GetItem(itemUid);
+
+            switch (shopItem.TokenType)
+            {
+                case CurrencyType.Meso:
+                    session.Player.Wallet.Meso.Modify(-(shopItem.Price * quantity));
+                    break;
+                case CurrencyType.ValorToken:
+                    session.Player.Wallet.ValorToken.Modify(-(shopItem.Price * quantity));
+                    break;
+                case CurrencyType.Treva:
+                    session.Player.Wallet.Treva.Modify(-(shopItem.Price * quantity));
+                    break;
+                case CurrencyType.Rue:
+                    session.Player.Wallet.Rue.Modify(-(shopItem.Price * quantity));
+                    break;
+                case CurrencyType.HaviFruit:
+                    session.Player.Wallet.HaviFruit.Modify(-(shopItem.Price * quantity));
+                    break;
+                case CurrencyType.Meret:
+                case CurrencyType.GameMeret:
+                case CurrencyType.EventMeret:
+                    session.Player.Wallet.Meret.Modify(-(shopItem.Price * quantity));
```

Comment:
The client checks for the **total** amount of merets the player has, so you should use Wallet function `RemoveMerets`